qmemman/meminfo-writer.c

```diff
@@ -19,41 +20,52 @@ int usr1_received;
 const char *parse(const char *meminfo_buf, const char* dom_current_buf)
 {
 	const char *ptr = meminfo_buf;
-	static char outbuf[4096];
+	static char outbuf[21];
 	long long val;
 	int len;
 	int ret;
 	long long MemTotal = 0, MemFree = 0, Buffers = 0, Cached = 0, SwapTotal =
 	    0, SwapFree = 0;
-	unsigned long long key;
 	long long used_mem, used_mem_diff;
 	int nitems = 0;
 
-	while (nitems != (1<<6)-1 || !*ptr) {
-		ret = sscanf(ptr, "%*s %lld kB\n%n", &val, &len);
-		if (ret < 1 || len < (int)sizeof (unsigned long long)) {
+	while (nitems != (1<<6)-1 && *ptr) {
+		ret = sscanf(ptr, "%*[A-Za-z]: %lld kB\n%n", &val, &len);
+		if (ret < 1 || len < 10) {
 			ptr += len;
 			continue;
 		}
-		key = *(unsigned long long *) ptr;
-		if (key == *(unsigned long long *) "MemTotal:") {
-			MemTotal = val;
-			nitems |= 1;
-		} else if (key == *(unsigned long long *) "MemFree:") {
-			MemFree = val;
-			nitems |= 2;
-		} else if (key == *(unsigned long long *) "Buffers:") {
-			Buffers = val;
-			nitems |= 4;
-		} else if (key == *(unsigned long long *) "Cached:  ") {
-			Cached = val;
-			nitems |= 8;
-		} else if (key == *(unsigned long long *) "SwapTotal:") {
-			SwapTotal = val;
-			nitems |= 16;
-		} else if (key == *(unsigned long long *) "SwapFree:") {
-			SwapFree = val;
-			nitems |= 32;
+		switch (*ptr) {
```

Comment:
This function is executed at least 10 times per second, in _every_ VM. Can you verify that your change do not make it any slower than it was? For example extract it into a tight loop and measure time for 10000 iterations or such.